File: packages/react-native-vision-camera/ios/Hybrid Objects/Recording/HybridVideoRecorder.swift
///
/// HybridVideoRecorder.swift
/// VisionCamera
/// Copyright © 2025 Marc Rousavy @ Margelo
///

import AVFoundation
import Foundation
import NitroModules

class HybridVideoRecorder: HybridRecorderSpec {
  private let videoOutput: AVCaptureMovieFileOutput
  private let queue: DispatchQueue
  private let fileURL: URL
  private let maxRecordedDuration: CMTime
  private let maxRecordedFileSize: Int64
  private var isCancelled = false

  init(
    videoOutput: AVCaptureMovieFileOutput,
    queue: DispatchQueue,
    fileType: RecorderFileType,
    settings: RecorderSettings
  ) throws {
    self.videoOutput = videoOutput
    self.queue = queue
    self.fileURL = try URL.createTempURL(fileType: fileType.toUTType())
    // `CMTime.invalid` / `0` mean "no limit" for `AVCaptureMovieFileOutput`.
    self.maxRecordedDuration =
      settings.maxDuration.map { CMTime(seconds: $0, preferredTimescale: 600) } ?? .invalid
    self.maxRecordedFileSize = settings.maxFileSize.map { Int64($0) } ?? 0
    super.init()
  }

  var isRecording: Bool {
    return videoOutput.isRecording
  }
  var isPaused: Bool {
    guard #available(iOS 18.0, *) else {
      return false
    }
    return videoOutput.isRecordingPaused
  }
  var recordedDuration: Double {
    return videoOutput.recordedDuration.seconds
  }
  var recordedFileSize: Double {
    return Double(videoOutput.recordedFileSize)
  }
  var filePath: String {
    return fileURL.absoluteString
  }

  func startRecording(
    onRecordingFinished: @escaping (_ filePath: String) -> Void,
    onRecordingError: @escaping (_ error: Error) -> Void,
    onRecordingPaused: (() -> Void)?,
    onRecordingResumed: (() -> Void)?
  ) throws -> Promise<Void> {
    let promise = Promise<Void>()
    var isResolved = false

    queue.async {
      // On our serial queue, make sure we are connected to the Camera
      guard self.videoOutput.connection(with: .video) != nil else {
        let error = RuntimeError.error(
          withMessage: "VideoOutput is not yet connected to the CameraSession!")
        promise.reject(withError: error)
        return
      }
      // On our serial queue, make sure we are currently not recoring
      guard !self.videoOutput.isRecording else {
        let error = RuntimeError.error(withMessage: "Active recording already in progress!")
        promise.reject(withError: error)
        return
      }

      // Prepare the delegate to resolve the Promise and notify listeners
      let delegate = VideoDelegate(
        onRecordingStarted: {
          // Recording started!
          promise.resolve()
          isResolved = true
        },
        onRecordingPaused: onRecordingPaused,
        onRecordingResumed: onRecordingResumed,
        onRecordingFinished: { url in
          if self.isCancelled {
            // Recording was cancelled - delete the file
            try? FileManager.default.removeItem(at: url)
            return
          }
          // Recording finished!
          onRecordingFinished(url.absoluteString)
        },
        onRecordingError: { error in
          if !isResolved {
            // We failed to even start the video recording!
            promise.reject(withError: error)
            isResolved = true
          } else {
            // An error occurred later on
            onRecordingError(error)
          }
        })

      // Apply recording limits. `AVCaptureMovieFileOutput` delivers the file
      // via the success path when either limit is reached (see `VideoDelegate`).
      self.videoOutput.maxRecordedDuration = self.maxRecordedDuration
      self.videoOutput.maxRecordedFileSize = self.maxRecordedFileSize

      // Start recording!
      self.videoOutput.startRecording(to: self.fileURL, recordingDelegate: delegate)
    }
    return promise
  }

  func stopRecording() throws -> Promise<Void> {
    return Promise.parallel(queue) {
      self.videoOutput.stopRecording()
    }
  }

  func pauseRecording() throws -> Promise<Void> {
    return Promise.parallel(queue) {
      guard #available(iOS 18.0, *) else {
        throw RuntimeError.error(
          withMessage: "`pauseRecording()` is only available on iOS 18 or above!")
      }
      self.videoOutput.pauseRecording()
    }
  }

  func resumeRecording() throws -> Promise<Void> {
    return Promise.parallel(queue) {
      guard #available(iOS 18.0, *) else {
        throw RuntimeError.error(
          withMessage: "`resumeRecording()` is only available on iOS 18 or above!")
      }
      self.videoOutput.resumeRecording()
    }
  }

  func cancelRecording() throws -> Promise<Void> {
    return Promise.parallel(queue) {
      guard self.videoOutput.isRecording else {
        throw RuntimeError.error(withMessage: "Not currently recording!")
      }
      // Mark as cancelled so the onRecordingFinished callback
      // deletes the file instead of notifying the caller
      self.isCancelled = true
      self.videoOutput.stopRecording()
    }
  }
}

File: packages/react-native-vision-camera/android/src/main/java/com/margelo/nitro/camera/hybrids/recording/HybridVideoRecorder.kt
package com.margelo.nitro.camera.hybrids.recording

import androidx.camera.video.PendingRecording
import androidx.camera.video.Recording
import androidx.camera.video.VideoRecordEvent
import com.margelo.nitro.camera.HybridRecorderSpec
import com.margelo.nitro.camera.extensions.VideoRecorderError
import com.margelo.nitro.camera.extensions.parallel
import com.margelo.nitro.camera.utils.IdentifiableExecutor
import com.margelo.nitro.core.Promise
import com.margelo.nitro.core.resolve
import java.io.File

class HybridVideoRecorder(
  private val pendingRecording: PendingRecording,
  private val file: File,
) : HybridRecorderSpec() {
  private val executor = IdentifiableExecutor("com.margelo.nitro.recorder")
  private var recording: Recording? = null
  private var isCancelled = false

  override val isRecording: Boolean
    get() = recording != null
  override var isPaused: Boolean = false
  override var recordedDuration: Double = 0.0
  override var recordedFileSize: Double = 0.0

  override val filePath: String
    get() = file.absolutePath

  override fun startRecording(
    onRecordingFinished: (String) -> Unit,
    onRecordingError: (Throwable) -> Unit,
    onRecordingPaused: (() -> Unit)?,
    onRecordingResumed: (() -> Unit)?,
  ): Promise<Unit> {
    var didResolve = false
    val promise = Promise<Unit>()

    executor.execute {
      // On our serial executor, make sure we are not currently recording
      if (this.recording != null) {
        val error = Error("Active recording already in progress!")
        promise.reject(error)
        return@execute
      }

      // Reset State
      this.isPaused = false
      this.recordedDuration = 0.0
      this.recordedFileSize = 0.0

      // Start recording
      this.recording =
        pendingRecording.start(executor) { event ->
          when (event) {
            is VideoRecordEvent.Start -> {
              promise.resolve()
              didResolve = true
            }
            is VideoRecordEvent.Pause -> onRecordingPaused?.invoke()
            is VideoRecordEvent.Resume -> onRecordingResumed?.invoke()
            is VideoRecordEvent.Finalize -> {
              // Recording is over - drop the Recording reference so `isRecording`
              // flips to false even when the finalize was auto-triggered (e.g. by
              // reaching `maxDuration` / `maxFileSize`) instead of via `stopRecording()`.
              this.recording = null
              if (isCancelled) {
                // Recording was cancelled - delete the file and don't notify
                isCancelled = false
                file.delete()
                return@start
              }
              when (event.error) {
                VideoRecordEvent.Finalize.ERROR_NONE,
                VideoRecordEvent.Finalize.ERROR_DURATION_LIMIT_REACHED,
                VideoRecordEvent.Finalize.ERROR_FILE_SIZE_LIMIT_REACHED,
                -> {
                  // Recording finished successfully - either no error, or file/duration limit reached:
                  val outputUri = event.outputResults.outputUri
                  onRecordingFinished(outputUri.toString())
                }
                else -> {
                  // We have an error, either during capture or even while starting:
                  val error = VideoRecorderError(event.error, event.cause)
                  if (!didResolve) {
                    // We didn't even start the Recording yet! Reject promise
                    promise.reject(error)
                    didResolve = true
                  } else {
                    // Unknown error while recording
                    onRecordingError(error)
                  }
                }
              }
            }
            is VideoRecordEvent.Status -> {
              this.recordedDuration = event.recordingStats.recordedDurationNanos / 1_000_000_000.0
              this.recordedFileSize = event.recordingStats.numBytesRecorded.toDouble()
            }
          }
        }
    }

    return promise
  }

  override fun stopRecording(): Promise<Unit> {
    return Promise.parallel(executor) {
      val recording = recording ?: throw Error("Not currently recording!")
      recording.stop()
      this.isPaused = false
      this.recording = null
      this.recordedDuration = 0.0
      this.recordedFileSize = 0.0
    }
  }

  override fun pauseRecording(): Promise<Unit> {
    return Promise.parallel(executor) {
      val recording = recording ?: throw Error("Not currently recording!")
      recording.pause()
      this.isPaused = true
    }
  }

  override fun resumeRecording(): Promise<Unit> {
    return Promise.parallel(executor) {
      val recording = recording ?: throw Error("Not currently recording!")
      recording.resume()
      this.isPaused = false
    }
  }

  override fun cancelRecording(): Promise<Unit> {
    return Promise.parallel(executor) {
      val recording = recording ?: throw Error("Not currently recording!")
      // Mark as cancelled so the Finalize event deletes
      // the file instead of notifying the caller
      this.isCancelled = true
      recording.close()
      this.isPaused = false
      this.recording = null
      this.recordedDuration = 0.0
      this.recordedFileSize = 0.0
    }
  }
}
